README.md
<!--
#
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
#  KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
#
-->

# Ripple Emulator(tm)

A browser based, platform agnostic mobile application development and testing tool.

## Build Requirements

* Common
  * `nodejs` (`v0.12.0` or higher)
  * `npm` (`v2.5.1` or higher)
  * `python` (`v2.7.3` recommended, `v3.x.x` is __not__ supported)
    * [Windows download](http://www.python.org/download/releases/2.7.3#download)

* OSX / Linux
    * `make`
    * A C/C++ compiler toolchain, e.g. GCC

* Windows
  * Visual Studio 2010
    * The setup instructions can be found [here](https://github.com/brianmcd/contextify/wiki/Windows-Installation-Guide)
  * [msysgit](http://msysgit.github.io/)
    * Ripple uses [Bower](http://bower.io/) for js libraries managing. In order to use Bower on Windows, [msysgit](http://msysgit.github.io/) must be installed in a proper way - see Bower's [README.md](https://github.com/bower/bower#windows-users)

## Getting Started

If you plan to dive into the source, be sure to check out the [HACKING](https://github.com/ripple-emulator/ripple/blob/master/HACKING.md) file.

To get started, you need to setup a few things, first- run (in the project root):

    ./configure

This script will pull down the needed npm packages and initialize the submodules.

## Build Commands

    jake

This will build ripple to the `pkg/` folder. In that folder there are various targets that can be used.

    jake -T

This will describe all the available commands for building and running the tests, etc.

## Running Inside Other Web Browsers

Ripple is (by-design) browser agnostic, and _should_ be able to run inside any web browser.

If you want to run it inside other browsers, you will need to use the `pkg/hosted` target, paired with the CLI's `emulate` command.

Ex (using the NPM package):

    ripple emulate --path to/my/app

    # or

    ripple emulate --remote http://remote-site.com

Then navigating to (your app's html file):

    http://localhost:PORT/index.html?enableripple=true

## CLI & NPM Package

There is a command line interface that can be paired with the client (UI).

It can be used for various things, such as statically hosting an application, and running a local (cross origin) XHR proxy.

To install:

    npm install -g ripple-emulator

This will install a global script called `ripple`. To see usage, run:

    ripple help

## Contributing

If you like the project, and want to contribute code, please issue a pull request (on [GitHub](https://github.com/ripple-emulator/ripple/pulls)) into the `master` branch.

## Code Guidelines

* 4 spaces per editor tab.
* `jake lint`, no new lint errors introduced.
* All unit tests are green.

## Reference Material &amp; Community

* [Project Site](https://github.com/ripple-emulator)


doc/CHANGELOG.md
<!--
#
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
#  KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
#
-->

## v0.9.36 - May 4, 2016

* Fix Angular 2 issue with XMLHttpRequest: [0973622e](https://github.com/ripple-emulator/ripple/commit/0973622e)

## v0.9.35 - March 11, 2016

* README was missing from package

## v0.9.34 - March 11, 2016

* Don't display "I Haz Cheeseburger" dialog for status bar plugin: [4aec49f5](https://github.com/ripple-emulator/ripple/commit/4aec49f5)

* Don't display "I Haz Cheeseburger" dialog for ionic keyboard plugin: [d304bf34](https://github.com/ripple-emulator/ripple/commit/d304bf34)

* Better unhandled exec call response when no success or fail callback: [6cb5b65f](https://github.com/ripple-emulator/ripple/commit/6cb5b65f)

## v0.9.33 - February 12, 2016

Ripple has moved to it's own stand-alone github repo at <https://github.com/ripple-emulator/ripple>. Issues and requests should be reported at <https://github.com/ripple-emulator/ripple/issues>.

* [Issue #9](https://github.com/ripple-emulator/ripple/issues/9): Added handling for `Media.MessageChannel` exec call: [2486d802](https://github.com/ripple-emulator/ripple/commit/2486d802)

* [Issue #8](https://github.com/ripple-emulator/ripple/issues/8): Work around crash with Media plugin introduced by bug in Chrome 48: [d9e4866b](https://github.com/ripple-emulator/ripple/commit/d9e4866b)

* [Issue #7](https://github.com/ripple-emulator/ripple/issues/7): Fix geolocation issues introduced by bug in in Chrome 48: [d9e4866b](https://github.com/ripple-emulator/ripple/commit/d9e4866b)

* [Issue #6](https://github.com/ripple-emulator/ripple/issues/6): Added handling for `Geolocation.getPermission` exec call on Android: [70c2ff33](https://github.com/ripple-emulator/ripple/commit/70c2ff33)

* Add support for Insomnia plugin ([cordova-plugin-insomnia](https://www.npmjs.com/package/cordova-plugin-insomnia)): [65c1034a](https://github.com/ripple-emulator/ripple/commit/65c1034a)

* ScanditSDK fixes: [e2c2f848](https://github.com/ripple-emulator/ripple/commit/e2c2f848) and [6d8bfcff](https://github.com/ripple-emulator/ripple/commit/6d8bfcff)

* [RIPPLE-99](https://issues.apache.org/jira/browse/RIPPLE-99) Updated emulation for Vibration API: [7578c471](https://github.com/ripple-emulator/ripple/commit/7578c471) 

* [RIPPLE-100](https://issues.apache.org/jira/browse/RIPPLE-100) Fixed battery status event emulation: [7578c471](https://github.com/ripple-emulator/ripple/commit/7578c471)

## v0.9.32 - August 16, 2015

* [RIPPLE-87](https://issues.apache.org/jira/browse/RIPPLE-87) Ripple can't emulate multi-page applications:
  <https://github.com/apache/incubator-ripple/pull/71>

* [RIPPLE-95](https://issues.apache.org/jira/browse/RIPPLE-95) Emulate command validates options incorrectly (regression):
  <https://github.com/apache/incubator-ripple/pull/70>

* There's no such thing as Earl *Gray* Tea! <https://github.com/apache/incubator-ripple/pull/69>

## v0.9.31 - July 23, 2015

* [RIPPLE-88](https://issues.apache.org/jira/browse/RIPPLE-88) Error in Cordova app with device motion plugin when running under Ripple:
  <https://github.com/apache/incubator-ripple/pull/62>

* [RIPPLE-89](https://issues.apache.org/jira/browse/RIPPLE-89) Battery Status panel throws an exception if app does not include cordova.js:
  <https://github.com/apache/incubator-ripple/pull/61>

* [RIPPLE-85](https://issues.apache.org/jira/browse/RIPPLE-85) Button index is incremented twice in notifications:
  <https://github.com/apache/incubator-ripple/pull/55>

* [RIPPLE-83](https://issues.apache.org/jira/browse/RIPPLE-83) Move where we call Cordova `prepare` so we prepare for the correct platform:
  <https://github.com/apache/incubator-ripple/pull/52>

* [RIPPLE-91](https://issues.apache.org/jira/browse/RIPPLE-91) Fix gpx file implementation in Geolocation panel:
  <https://github.com/apache/incubator-ripple/pull/63>

* Improve layout and display of units in Accelerometer panel:
  <https://github.com/apache/incubator-ripple/pull/66>

* [RIPPLE-93](https://issues.apache.org/jira/browse/RIPPLE-93) Cordova `prepare` doesn't work if cwd isn't in Cordova project.
  Also, validate command line options, and move where we get a reference to Cordova `prepare` to a later point (so it is after
  we validate command line options):
  <https://github.com/apache/incubator-ripple/pull/65>

* [RIPPLE-94](https://issues.apache.org/jira/browse/RIPPLE-94) Fix issues with emulation of Console and Contacts plugins identified by running Mobile Spec: 
  <https://github.com/apache/incubator-ripple/pull/68>

## v0.9.30 - May 27, 2015

* RIPPLE-86 Changes to allow Ripple to launch in Internet Explorer and Edge browsers:
  <https://github.com/apache/incubator-ripple/pull/60>

* Further cleanup in LICENSE file:
  <https://github.com/apache/incubator-ripple/pull/57> and
  <https://github.com/apache/incubator-ripple/pull/58>

* README.md updates:
  <https://github.com/apache/incubator-ripple/pull/54> and
  <https://github.com/apache/incubator-ripple/pull/53>

## v0.9.29 - May 12, 2015

* RIPPLE-74 Ripple was not passing geolocation info to the app
  <https://github.com/apache/incubator-ripple/pull/51>

* Updated versions of jsdom and browserify so installs with Node 0.11.0 and above.

lib/client/platform/cordova/3.0.0/spec.js
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
module.exports = {
    id: "cordova",
    version: "3.0.0",
    name: "Apache Cordova",
    type: "platform",
    nativeMethods: {},

    config: ripple('platform/cordova/3.0.0/spec/config'),
    device: ripple('platform/cordova/2.0.0/spec/device'),
    ui: ripple('platform/cordova/2.0.0/spec/ui'),
    events: ripple('platform/cordova/2.0.0/spec/events'),

    initialize: function (win) {
        var honeypot = ripple('honeypot'),
            devices =ripple('devices'),
            device = devices.getCurrentDevice(),
            bridge = ripple('platform/cordova/2.0.0/bridge'),
            cordova,
            get = function () {
                return cordova;
            },
            set = function (orig) {
                if (cordova) return;

                cordova = orig;

                cordova.define.remove("cordova/exec");
                cordova.define("cordova/exec", function (require, exports, module) {
                    module.exports = bridge.exec;
                    module.exports.init = function(){cordova.require('cordova/channel').onNativeReady.fire();}; // RIPPLE-69 adding cordova-3.6.x support; RIPPLE-87 init method overwritting so that its closure points to the current cordova object
                });
            };

        bridge.add("PluginManager", ripple('platform/cordova/3.0.0/bridge/PluginManager'));
        bridge.add("SplashScreen", ripple('platform/cordova/3.0.0/bridge/splashscreen'));
        bridge.add("InAppBrowser", ripple('platform/cordova/3.0.0/bridge/inappbrowser'));
        bridge.add("Vibration", ripple('platform/cordova/3.0.0/bridge/vibration'));

        // RIPPLE-79 cordova-android 4.0.0 renames 'App' service to 'CoreAndroid'. We'll leave
        // platform/cordova/2.0.0/bridge/app in place for backwards compatibility, and add CoreAndroid
        // here for the future.
        bridge.add("CoreAndroid", ripple('platform/cordova/3.0.0/bridge/coreandroid'));

        /** external plugins **/
        // scandit barcode scanner @see https://github.com/Scandit/BarcodeScannerPlugin
        bridge.add("ScanditSDK", ripple('platform/cordova/3.0.0/bridge/scanditsdk'));

        // Eddy Verbruggen's Insomnia-PhoneGap-Plugin (https://github.com/EddyVerbruggen/Insomnia-PhoneGap-Plugin)
        bridge.add("Insomnia", ripple('platform/cordova/3.0.0/bridge/insomnia'));

        // RIPPLE-71 BarcodeScaner plugin support @see https://github.com/phonegap-build/BarcodeScanner
        bridge.add("BarcodeScanner", ripple('platform/cordova/3.0.0/bridge/barcodescanner'));

        bridge.add("NfcPlugin", ripple('platform/cordova/3.0.0/bridge/nfcdummy'));

        // Don't display I Haz Cheeseburger?!?! dialog for ionic keyboard plugin init call
        bridge.add("Keyboard", ripple('platform/cordova/3.0.0/bridge/keyboard'));

        // Don't display I Haz Cheeseburger?!?! dialog for statusbar plugin _ready call
        bridge.add("StatusBar", ripple('platform/cordova/3.0.0/bridge/statusbar'));

        honeypot.monitor(win, "cordova").andRun(get, set);

        //HACK: BlackBerry does vibration different
        if (device.manufacturer === "BlackBerry") {
            navigator.vibrate = function (ms) {
                ripple('platform/cordova/3.0.0/bridge/vibration').vibrate(null, null, [ms]);
            };
        }
    },

    objects: {
        MediaError: {
            path: "cordova/2.0.0/MediaError"
        },
        Acceleration: {
            path: "w3c/1.0/Acceleration"
        },
        Coordinates: {
            path: "w3c/1.0/Coordinates"
        },
        Position: {
            path: "w3c/1.0/Position"
        },
        PositionError: {
            path: "w3c/1.0/PositionError"
        },
        navigator: {
            path: "w3c/1.0/navigator",
            children: {
                geolocation: {
                    path: "w3c/1.0/geolocation"
                }
            }
        },
        org: {
            children: {
                apache: {
                    children: {
                        cordova: {
                            children: {
                                Logger: {
                                    path: "cordova/2.0.0/logger"
                                },
                                JavaPluginManager: {
                                    path: "cordova/2.0.0/JavaPluginManager"
                                }
                            }
                        }
                    }
                }
            }
        }
    }
};
